packages/python/plotly/plotly/matplotlylib/renderer.py

```diff
@@ -312,10 +312,84 @@ def draw_bar(self, coll):
                 "assuming data redundancy, not plotting."
             )
 
+    def draw_legend_shapes(self, mode, shape, **props):
```

Comment:
I'm not sure I understand what this function does... could you please explain or provide an example pair of matplotlib/plotly figures that uses this system?

Reply:
Without this function which is called further down on line 583 when coordinates are "axes" (plotly only handles "data"), the plots legends have no markers or lines, just text. This is why we get the warning: "Bummer! Plotly can currently only draw Line2D objects from matplotlib that are in 'data' coordinates!"

See example notebook here:

https://colab.research.google.com/drive/1-MNmqy_DeDxw8yPX0U7M9E8AEMtfIobP?usp=sharing

Matplotlib plot:
![index](https://user-images.githubusercontent.com/1105325/120505365-e6d08980-c392-11eb-893c-84c260ebc076.png)


Plotly version:
![newplot(2)](https://user-images.githubusercontent.com/1105325/120504972-90634b00-c392-11eb-966e-cc1fbc8aa389.png)


When coordinates == axes I call my new function that draws the shapes. go.layout.Shape cant handle circles, so I handle this special case, handle lines, and everything else i pass to go.layout.shape.

Result:
![newplot(3)](https://user-images.githubusercontent.com/1105325/120505930-75450b00-c393-11eb-8ae5-0e0b80d070d1.png)

It still is left with one warning: "I found a path object that I don't think is part of a bar chart. Ignoring."

This is due to the bounding box that in matplotlib is (now) a fancy box with curved corners. This would have to be addressed in a future PR.